Improve start

richardkiss opened the pull request.

richardkiss committed:
Checkpoint.

richardkiss committed:
Add start_daemon_client.

richardkiss committed:
First crack at `chia daemon`.

richardkiss committed:
Improve UI.

richardkiss committed:
More improvements.

richardkiss committed:
Fix stop problem with groups.

richardkiss committed:
Simplify stop.

richardkiss committed:
Having problems with argparse.

richardkiss committed:
Try hard to prevent multiple chia daemon instances from running.

richardkiss committed:
Add a semaphore.

richardkiss committed:
Use UNIX sockets on platforms that have them.

richardkiss committed:
Fix problem with UNIX sockets.

richardkiss committed:
Merge remote-tracking branch 'origin/dev' into improve_start

richardkiss committed:
Remove lockfile on Windows.

richardkiss committed:
API server is now an aiter.

richardkiss committed:
Don't print anything when launching chia_farmer.

richardkiss committed:
Validate services before starting.

richardkiss committed:
Add comments.

richardkiss committed:
Close lockfile. Deal with subprocess launching errors better.

richardkiss committed:
Merge remote-tracking branch 'origin/dev' into improve_start

richardkiss committed:
Move service_groups file.

richardkiss committed:
Remove pid_path_killed if it exists before renaming.

richardkiss commented:
@hoffmang9 This is ready to start trying to see how broken it is. Please try it on Windows and UNIX platforms.

richardkiss commented:
Features include:

- `chia run_daemon` to create a long-lived daemon instance that launches all subprocesses
- in `chia stop -d all` the `-d` will also kill the daemon
- on UNIX, use a unix socket in the `run` directory
- on Windows, uses a TCP listen socket, trying a bunch starting at 60191 (so usually 60191)
- uses lockfiles to make it difficult/impossible to launch multiple daemons

The code uses a lot of code pulled from ledger-sim to make request/responses simpler. I plan to port it to something that looks much more like what we have elsewhere to unify the architecture.

richardkiss committed:
Bind to localhost.

richardkiss committed:
Fix flake8.

mariano54 commented:
The daemon is meant to be called only internally right? For shutting down, starting, and accessing the node RPCs. It's not used by external nodes, and it's not used for nodes to communicate with each other within this machine. If so, why have a custom protocol and use CBOR here? Cbor is only used for the chia protocol (farmer protocol, full node, etc), and these services connect to each other directly, not through the daemon.

Isn't it simpler to expose an HTTP and websocket server and anyone can easily build an rpc client using an existing HTTP client library, and call the correct methods? I'm not sure why we need to use unix sockets and create a custom protocol here. 

Under the unix socket or tcp socket approach, making a client is much more difficult right? For example here is the client for the electron app for the full node: https://github.com/Chia-Network/chia-blockchain/blob/master/electron-ui/full_node/full_node_rpc_client.js. Making a client for a custom cbor protocol on a unix socket seems much more error prone and complicated.

Also having an http server means a uniform implementation for windows and unix

richardkiss commented:
I used code from `ledger_sim` for expediency, but yes, I'm with you, it was only intended to be temporary. I actually would like to use http for request/response and websockets for peer-to-peer. In fact, I mentioned in a daily call that you weren't on that I'm going to investigate replacing cbor framing with websockets framing since we already have a websockets dependency. This would an incompatible change, so probably best to do at a hard fork (although I guess we could support both transports for a while), but then it would be possible to create first-class web clients.

richardkiss committed:
Fix some PEP8.

richardkiss committed:
First crack at http RPC.

richardkiss committed:
Move client over to http RPC.

richardkiss committed:
Checkpoint. Remove proxy API.

richardkiss committed:
Use UNIX sockets.

richardkiss committed:
Merge remote-tracking branch 'origin/dev' into improve_start

richardkiss committed:
Remove obsolete file.

richardkiss committed:
Comments, minor improvements.

richardkiss committed:
Remove another file.

richardkiss committed:
LGTM tweaks.

richardkiss commented:
I've removed the custom protocol and replaced it with http RPC without much trouble. A small window has opened up which allow multiple daemons to be launched, so that window needs to be closed. I have to dig into `aiohttp` a bit deeper to do so (essentially, we can't close the lock file until the socket is listening). Exiting the daemon is also kind of gross and needs to be tidied up. Launching on Windows requires the virtual env to be in place before the daemon is launched, which is not ideal. But at least the `src.proxy` module is gone.

richardkiss commented:
@mariano54 @hoffmang9  Any comments?

richardkiss commented:
(We'd want to squash merge this mess if we're okay to merge it.)

mariano54 commented on src/daemon/server.py:
```diff
@@ -0,0 +1,239 @@
+            log.info("sending kill signal to %s", service_name)
+    r = process.wait()
+    log.info("process %s returned %d", process, r)
+    try:
```
What is the point of saving the pids? If we already have the process saved in the daemon

mariano54 replied:
Also the PID stuff was broken in windows right?

mariano54 commented on src/daemon/server.py:
```diff
@@ -0,0 +1,239 @@
+        services.clear()
+
+        # TODO: fix this hack
+        asyncio.get_event_loop().call_later(5, lambda *args: sys.exit(0))
```
Is this so that we can return the HTTP response?

richardkiss replied:
Yeah. Also, getting a clean exit can be tricky.

mariano54 commented on src/daemon/server.py:
```diff
@@ -0,0 +1,239 @@
+    """
+    try:
+        if fcntl:
+            f = open(lockfile, "w")
```
Did you try this on windows?

richardkiss replied:
The `fcntl` module doesn't exist on windows. That's why the conditional is there.

mariano54 reviewed (approved):
Looks much better to me. We should fix those race conditions if we can. Also we need to test that it works on all platforms, so we don't introduce bugs into master and then have to make an emergency fix.

hoffmang9 commented:
Good news so far on Win 10:
1. Start a farmer and sync up using Powershell "bg" method.
2. `chia stop farmer`
3. Minor downside - Windows thinks the job is still running but
4. Process stops correctly from our perspective
5. And I can restart it with the windows Jobs bg method and it takes right off.
6. Then I restarted without stopping.
7. And everything worked when I followed our "start farming again" instructions for Windows.

hoffmang9 commented:
This is ready to merge but there are conflicts in:
src/cmds/start.py
src/cmds/start_stop.py

@richardkiss - can you resolve those and then I'll merge?

hoffmang9 commented:
Did this get addressed yet or will you want to address it after merge?
https://github.com/Chia-Network/chia-blockchain/pull/217/checks?check_run_id=672936127

richardkiss committed:
Merge remote-tracking branch 'origin/dev' into improve_start
